Add Typescript annotations to Select, SelectItem and SelectItemGroup

jpsorensen opened the pull request:
Closes #12555 

Convert existing propType definitions to Typescript annotations on Select, SelectItem and SelectItemGroup.

#### Changelog

**Changed**

- Select.js, SelectItem.js and SelectItemGroup.js converted to tsx
- Removed label attribute from unit test

#### Testing / Reviewing

- In the packages/react directory, run yarn build, yarn storybook. In the storybook, navigate to Select and confirm that the component's behavior remains unchanged.
- Find usages of Select (along with SelectItem and SelectItemGroup) throughout the code base. Add them to the bottom of Select.tsx and confirm that they compile. (In this case, one did not and I eventually ended up changing the test - removing label. See notes below).

jpsorensen committed:
refactor(typescript): type annotations on Select*

Add Typescript annotations to Select, SelectItem and SelectItemGroup.

jpsorensen committed:
Merge branch 'main' into refactor/12555-select-typescript

jpsorensen commented on packages/react/src/components/Select/__tests__/Select-test.js:
```diff
@@ -174,7 +174,6 @@ describe('Select', () => {
       render(
         <Select
           id="select-1"
-          label="Select label"
```
This does not pass compilation, as there's no label attribute on Select.

Spent some time digging around and eventually concluded it should not be there, but happy to change this if someone else knows something I missed.

There's no explicitly defined label attribute. It does pass its ...other attributes on to the <select> element, but according to my searching (both looking at the Typescript definition of select and looking at docs for <select> online, I don't think it has a label attribute - it can be associated with a <label> element, but as far as I could tell setting a label= attribute is not a shorthand for an actual label element.

andreancardona replied:
I'll defer to @tay1orjones on this one 👀

tw15egan replied:
I believe it should be using the `labelText` prop, not `label`

jpsorensen replied:
Great, thanks for weighing in! Updated to use `labelText` instead of `label`.

jpsorensen committed:
fix(lint): removed unused imports

Left SelectItem and SelectItemGroup in the imports after using them for
testing. Removed.

jpsorensen committed:
fix(Select-test): update label to labelText

Previously removed label as it was not a property of Select.
On recommendation of maintainers reintroduced using labelText.

andreancardona committed:
Merge branch 'main' into refactor/12555-select-typescript

andreancardona commented on packages/react/src/components/Select/Select.tsx:
```diff
@@ -18,6 +25,109 @@ import { useFeatureFlag } from '../FeatureFlags';
+  invalidText?: ReactNode;
+
+  /**
+   * Provide label text to be read by screen readers when interacting with the
```
```suggestion
   * Provide label text to be read by screen readers when interacting with the control.
```

andreancardona commented on packages/react/src/components/Select/Select.tsx:
```diff
@@ -18,6 +25,109 @@ import { useFeatureFlag } from '../FeatureFlags';
+
+  /**
+   * Provide label text to be read by screen readers when interacting with the
+   * control
```
```suggestion
```

andreancardona commented on packages/react/src/components/Select/Select.tsx:
```diff
@@ -18,6 +25,109 @@ import { useFeatureFlag } from '../FeatureFlags';
+   * `true` to use the light version. For use on $ui-01 backgrounds only.
+   * Don't use this to make tile background color same as container background color.
+   *
+   * @deprecated The `light` prop for `Select` is no longer needed and has
```
```suggestion
   * @deprecated The `light` prop for `Select` is no longer needed and has been deprecated in v11 in favor of the new `Layer` component.
```

andreancardona commented on packages/react/src/components/Select/Select.tsx:
```diff
@@ -18,6 +25,109 @@ import { useFeatureFlag } from '../FeatureFlags';
+   * Don't use this to make tile background color same as container background color.
+   *
+   * @deprecated The `light` prop for `Select` is no longer needed and has
+   *    been deprecated in v11 in favor of the new `Layer` component. It
```
```suggestion
   * It will be moved in the next major release.
```

andreancardona commented on packages/react/src/components/Select/Select.tsx:
```diff
@@ -18,6 +25,109 @@ import { useFeatureFlag } from '../FeatureFlags';
+   *
+   * @deprecated The `light` prop for `Select` is no longer needed and has
+   *    been deprecated in v11 in favor of the new `Layer` component. It
+   *    will be moved in the next major release.
```
```suggestion
```

andreancardona commented on packages/react/src/components/Select/Select.tsx:
```diff
@@ -18,6 +25,109 @@ import { useFeatureFlag } from '../FeatureFlags';
+  noLabel?: boolean;
+
+  /**
+   * Provide an optional `onChange` hook that is called each time the value of
```
```suggestion
   * Provide an optional `onChange` hook that is called each time the value of the underlying `<input>` changes.
```

andreancardona commented on packages/react/src/components/Select/Select.tsx:
```diff
@@ -18,6 +25,109 @@ import { useFeatureFlag } from '../FeatureFlags';
+
+  /**
+   * Provide an optional `onChange` hook that is called each time the value of
+   * the underlying `<input>` changes
```
```suggestion
```

jpsorensen committed:
Apply suggestions from code review

Co-authored-by: Andrea N. Cardona <cardona.n.andrea@gmail.com>

jpsorensen committed:
Follow up to change suggestions

jpsorensen committed:
Merge branch 'main' into refactor/12555-select-typescript

andreancardona committed:
Merge branch 'main' into refactor/12555-select-typescript

tw15egan reviewed (approved):
LGTM 👍 ✅

kodiakhq[bot] committed:
Merge branch 'main' into refactor/12555-select-typescript
